pulsar-sql/presto-pulsar/src/main/java/org/apache/pulsar/sql/presto/decoder/protobufnative/PulsarProtobufNativeColumnDecoder.java

```diff
@@ -0,0 +1,386 @@
+                format("cannot decode object of '%s' as '%s' for column '%s'",
+                        value.getClass(), type, columnName));
+    }
+
+    private static Block serializeObject(BlockBuilder builder, Object value, Type type, String columnName) {
+        if (type instanceof ArrayType) {
+            return serializeList(builder, value, type, columnName);
+        }
+        if (type instanceof MapType) {
+            return serializeMap(builder, value, type, columnName);
+        }
+        if (type instanceof RowType) {
+            return serializeRow(builder, value, type, columnName);
+        }
+        serializePrimitive(builder, value, type, columnName);
+        return null;
+    }
+
+    private static Block serializeList(BlockBuilder parentBlockBuilder, Object value, Type type, String columnName) {
+        if (value == null) {
+            checkState(parentBlockBuilder != null, "parentBlockBuilder is null");
+            parentBlockBuilder.appendNull();
+            return null;
+        }
+        List<?> list = (List<?>) value;
+        List<Type> typeParameters = type.getTypeParameters();
+        Type elementType = typeParameters.get(0);
+
+        BlockBuilder blockBuilder = elementType.createBlockBuilder(null, list.size());
+        for (Object element : list) {
+            serializeObject(blockBuilder, element, elementType, columnName);
+        }
+        if (parentBlockBuilder != null) {
+            type.writeObject(parentBlockBuilder, blockBuilder.build());
+            return null;
+        }
+        return blockBuilder.build();
+    }
+
+    private static void serializePrimitive(BlockBuilder blockBuilder, Object value, Type type, String columnName) {
+        requireNonNull(blockBuilder, "parent blockBuilder is null");
+
+        if (value == null) {
+            blockBuilder.appendNull();
+            return;
+        }
+
+        if (type instanceof BooleanType) {
+            type.writeBoolean(blockBuilder, (Boolean) value);
+            return;
+        }
+
+        if ((value instanceof Integer || value instanceof Long)
+                && (type instanceof BigintType || type instanceof IntegerType
+                || type instanceof SmallintType || type instanceof TinyintType)) {
+            type.writeLong(blockBuilder, ((Number) value).longValue());
+            return;
+        }
+
+        if (type instanceof DoubleType) {
+            type.writeDouble(blockBuilder, (Double) value);
+            return;
+        }
+
+        if (type instanceof RealType) {
+            type.writeLong(blockBuilder, floatToIntBits((Float) value));
+            return;
+        }
+
+        if (type instanceof VarcharType || type instanceof VarbinaryType) {
+            type.writeSlice(blockBuilder, getSlice(value, type, columnName));
+            return;
+        }
+
+        if (type instanceof TimestampType) {
+            type.writeLong(blockBuilder, (Long) value);
+            return;
```

Comment:
some static code are very similar to ```PulsarAvroColumnDecoder```, It seems that some should be deleted or unified in to one method

Reply:
Thank you for remind , `PulsarProtobufNativeRowDecoderFactory.parseProtobufPrestoType() ` will not return TimestampType , so this logic is redundancy , I deleted .
 Logic abstract may be a complex job ,  on current version i don't think this is a cost-effective optimize ?